File: Top200/19-RemoveNthFromEnd.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil {
		return head
	}

	dummy := &ListNode{Val: -1}
	dummy.Next = head

	fast := dummy
	for i := 0; i < n; i++ {
		fast = fast.Next
	}

	slow := dummy
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}

	slow.Next = slow.Next.Next

	return dummy.Next
}

File: Top200/19-RemoveNthFromEnd.java
package Top200;

class RemoveNthFromEnd {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        ListNode curr = dummy;
        curr.next = head;

        for (int i = 1; i <= n; i++) {
            curr = curr.next;
        }        

        while (curr.next != null) {
            curr = curr.next;
            prev = prev.next;
        }

        prev.next = prev.next.next;

        return dummy.next;
    }
    
}
